mypy_django_plugin/lib/helpers.py

```diff
@@ -69,12 +69,28 @@ def lookup_fully_qualified_sym(fullname: str, all_modules: Dict[str, MypyFile])
     else:
         module, cls_name = fullname.rsplit(".", 1)
 
-    module_file = all_modules.get(module)
-    if module_file is None:
-        return None
-    sym = module_file.names.get(cls_name)
+    module_file = None
+    while True:
+        module_file = all_modules.get(module)
+        if module_file:
+            break
+        else:
+            if "." not in module:
+                return None
+            module, parent_cls = module.rsplit(".", 1)
+            cls_name = ".".join([parent_cls, cls_name])
```

Comment:
I can't recall exactly where but I think mypy has utility that tells you what scope is currently being processed. Could we investigate if that could be useful/a possibility here? Instead of splitting and retrying on the fullname, potentially reimplementing what mypy has for us.

With that said, perhaps it could be up to the caller to be aware of class level scope and rsplit the fullname before passing it to `lookup_fully_qualified_sym`?

What would/could such an approach look like?

Reply:
I'll investigate this, makes more sense. It ended up being more hackish than I would have liked.

Reply:
I found [this](https://github.com/python/mypy/blob/486a51b6e49fff63c05b139185a560c85db7f7be/mypy/lookup.py#L13-L61) in the mypy codebase, which apparently could be a full replacement for the `lookup_fully_qualified_sym` function in this plugin, but I am having a bit of trouble integrating everything because I am to unable place a pdb breakpoint in the plugin code while running the tests. It just hangs there when it finds a `breakpoint()` or `pdb.set_trace()` call.

Is there any way I can setup my django-stubs installation to be able to debug tests?

Reply:
The `lookup_fully_qualified` you've found looks promising, but I'm wondering why it wasn't used in the first place. I would suspect, since they're named so similar, someone was aware of it when implementing `lookup_fully_qualified_sym`.

Although, I'd say our test suite should back up if `lookup_fully_qualified` is a valid replacement. So if the suite is green after a swap, everything is fine IMO.

IIRC I've not been able to use breakpoints either. Been resorting to the good ol' print debugging 🙂

Reply:
Ok, so I went deeper into the rabbit hole and rewrote the pull request. The version I mentioned from mypy is quite basic and does not support all of mypy_django_plugin use cases. This rewritten version should be clearer, easier to maintain than my previous one and now passes all tests. It's also probably a bit more performant, as it avoids unnecessary string joining and splitting.

I also found a way to use breakpoints with pytest_mypy_plugins. I will submit a pull request shortly to address that.

Reply:
> ... I also found a way to use breakpoints with pytest_mypy_plugins. I will submit a pull request shortly to address that.

Cool, sounds great

Reply:
This is the pull request: https://github.com/typeddjango/pytest-mypy-plugins/pull/108